### DISCUSSION TITLE
Coordinator vs. Optimizer Tasks

### DISCUSSION THREAD

**Discussion by keyseven123** [nebulastream/nebulastream#159] [Ideas]
This discussions should revolve around what we mean when we speak about an optimizer and a coordinator. There is another discussion closely related to this that discusses [a language frontend and an optimizer](https://github.com/nebulastream/nebulastream-public/discussions/47). 
The best outcome of this discussion is a clear definition and list of tasks (might be extended) of what tasks a coordinator have to do and what tasks an optimizer must do.
As this discussion evolves, we will update this comment here and add the agreed upon tasks to the coordinator or optimizer list.

Coordinator:
- *coordinator tasks*

Optimizer:
- *optimizer tasks*

**Comment by keyseven123** [comment]
# Handling lifecycle of a query
IMO, a coordinator should handle the lifecycle of a query. This entails:
- (re-)deploying a query 
- calling the optimizer and getting an optimized query plan back
- stopping queries once the coordinator receives a fail query request from one of the workers

**Reply by zeuchste** [reply]
Yes the general one could be here but as we have a client/server architecture there are also parts in the worker but I think this is not relevant here as you compare coordinator vs. optimizer. I would call this rather query lifetime management

**Comment by keyseven123** [comment]
# Handling (un-)registering of Worker (Topology Management)
IMO, a coordinator should handle the registration of workers and the topology. This entails having a gRPC or RestAPI that the worker can connect and disconnect from.

**Reply by zeuchste** [reply]
yes and also redirect if failure happens?

**Comment by keyseven123** [comment]
# Handling the Requests from Users
IMO, a coordinator should handle requests from users:
- Starting and stopping of queries
- Adding and removing of (logical) streams
- Changing permissions of users or of streams

**Comment by keyseven123** [comment]
# Parsing of queries and optimizing
IMO the only tasks an optimizer has to do is 
- parse a query
- optimize a specific query

It can store or access anything necessary for performing both tasks, e.g., topology or source catalog.

**Reply by adrianmichalke** [reply]
It think there might be more tasks specifically the re-optimization of existing queries: 
- re-optimize when topology changes
- re-optimize when a new query is issued (source/operator sharing)
- gather statistics
- query validation: i) syntax check, ii) name resolution (e.g. check if source names exist)

afaik, the optimization does not happen on the base of a individual query but on multiple queries together (query graph)

But its a cool starting point, I think we can come up with a clear interface between both components!

**Reply by zeuchste** [reply]
I think the parser is not part of the optimizer. I think the best definition of an optimizer is by its interface, it takes a valid query plan and it return a valid query plan. Everything else is not his job.

The maintenance of the catalogs as you mention is I think also the part of the coordinator
